

<!DOCTYPE html>
<html lang="en" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">
  <link rel="apple-touch-icon" sizes="76x76" href="/img/page_avatar.png">
  <link rel="icon" href="/img/page_avatar.png">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="Emmett Peng">
  <meta name="keywords" content="">
  
    <meta name="description" content="Project: Interdomain network analysis between prokaryotes and eukaryotes in the offshore area of Bohai sea2019.09 - 2020.09 Prokaryotic and eukaryotic microorganisms are major components of oceanic ec">
<meta property="og:type" content="website">
<meta property="og:title" content="Research Overview">
<meta property="og:url" content="https://emmettpeng.github.io/research/index.html">
<meta property="og:site_name" content="Emmett&#39;s Note">
<meta property="og:description" content="Project: Interdomain network analysis between prokaryotes and eukaryotes in the offshore area of Bohai sea2019.09 - 2020.09 Prokaryotic and eukaryotic microorganisms are major components of oceanic ec">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2025-11-17T14:55:51.000Z">
<meta property="article:modified_time" content="2025-11-17T15:07:10.235Z">
<meta property="article:author" content="Emmett Peng">
<meta name="twitter:card" content="summary_large_image">
  
  
  
  <title>Research Overview - Emmett&#39;s Note</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />





<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/font_1749284_hj8rtnfg7um.css">



<link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_lbnruvf0jn.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"emmettpeng.github.io","root":"/","version":"1.9.5-a","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"woyaola":null,"cnzz":null,"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0"></head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 60vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>emmett&#39;s notebook</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/">
                <i class="iconfont icon-home-fill"></i>
                <span>Home</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/publications/">
                <i class="iconfont icon-books-fill"></i>
                <span>Publications</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/research/">
                <i class="iconfont icon-exp-fill"></i>
                <span>Research</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/">
                <i class="iconfont icon-user-fill"></i>
                <span>About</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/links/">
                <i class="iconfont icon-link-fill"></i>
                <span>Links</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('/img/default.png') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="Research Overview"></span>
          
        </div>

        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      <div class="container nopadding-x-md">
        <div id="board"
          >
          
          <div class="container">
            <div class="row">
              <div class="col-12 col-md-10 m-auto">
                

<article class="page-content">
  <h2 id="Project-Interdomain-network-analysis-between-prokaryotes-and-eukaryotes-in-the-offshore-area-of-Bohai-sea"><a href="#Project-Interdomain-network-analysis-between-prokaryotes-and-eukaryotes-in-the-offshore-area-of-Bohai-sea" class="headerlink" title="Project: Interdomain network analysis between prokaryotes and eukaryotes in the offshore area of Bohai sea"></a>Project: Interdomain network analysis between prokaryotes and eukaryotes in the offshore area of Bohai sea</h2><p>2019.09 - 2020.09</p>
<p>Prokaryotic and eukaryotic microorganisms are major components of oceanic ecosystems, contributing significantly to nutrient cycling, primary productivity, and overall marine health. However, knowledge about their complex interactions remains scarce, particularly at the community level. In this study, we initially investigated the microbial community structure and prokaryote-eukaryote interactions in the offshore area of Changxing Island Monitoring Area, Bohai Sea. We utilized the interdomain ecological network (IDEN) analysis to visualize intricate connections among the microbial taxa, revealing that sediment-based microbial communities exhibit tighter connections and more complex interactions compared to those in seawater. Network centrality measures were used to identify keystone taxa, which revealed that the SAR supergroup was the dominant phyla in the IDEN, suggesting their pivotal role in microbial interactions. Moreover, beta-diversity (dissimilarity) analysis indicated that nearshore and far offshore samples displayed significant differences in community composition and prokaryote-eukaryote interactions, which could be influenced by factors like nutrient gradients and human activities. Through this research project, I gained a preliminary understanding of the study of microbial ecology and acquired foundational skills in research methods such as 16S rRNA gene amplicon analysis, molecular ecological network analysis, and conventional statistical analysis.</p>
<h2 id="Integrate-metabolic-complementarity-and-co-occurrence-into-microbial-interaction-modeling"><a href="#Integrate-metabolic-complementarity-and-co-occurrence-into-microbial-interaction-modeling" class="headerlink" title="Integrate metabolic complementarity and co-occurrence into microbial interaction modeling"></a>Integrate metabolic complementarity and co-occurrence into microbial interaction modeling</h2><p>2020.09 - 2023.12</p>
<p>The microbial communities are complex interconnected ecological communities that cross-feed, communicate, and coevolve, but these interactions are hard to observe directly. The co-occurrence patterns and metabolic models are two commonly used approaches to infer microbial interactions. However, the synthetical applications of both approaches were rarely brought up to resolve complicated microbial interactions in natural ecosystems. Here, we integrated metabolic complementarity and co-occurrence patterns through the random matrix theory (RMT) approach and further revealed the microbial interactions from a series of hot spring samples under a temperature gradient. Both network modeling demonstrated that the microbial interaction pattern becomes denser with temperature increases. The overlapped species interactions have been identified from these two network approaches, while metabolites such as amino acids, coenzyme A derivatives, and carbohydrates were essential for their metabolic exchange in the hot spring. We also observed a positive correlation between the exchange of basal metabolites in microbial species interactions and their genome size differences, suggesting metabolite exchange is a potential means for microorganisms with streamlined genomes to cooperate with others. For this analysis workflow, a publicly available pipeline has been integrated into our iNAP website (https:&#x2F;&#x2F; inap.denglab.org.cn) to support the measurement of metabolic complementarity and co-occurrence from metagenome-assembled genomes (MAGs) and RMT threshold determination.</p>
<p>Related publications:</p>
<p><a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-024-52532-x">Peng et al., <em>Nature Communications.</em> (2024)</a><br><a target="_blank" rel="noopener" href="https://onlinelibrary.wiley.com/doi/full/10.1002/imt2.235">Peng et al., <em>iMeta.</em> (2024)</a></p>
<h2 id="Project-Construct-microbial-interaction-networks-from-the-quorum-sensing-QS-perspective"><a href="#Project-Construct-microbial-interaction-networks-from-the-quorum-sensing-QS-perspective" class="headerlink" title="Project: Construct microbial interaction networks from the quorum sensing (QS) perspective"></a>Project: Construct microbial interaction networks from the quorum sensing (QS) perspective</h2><p>2024.01 - Present</p>
<p>Microbial interactions are often mediated by chemical signaling, among which quorum sensing (QS) plays a central role in coordinating behaviors across species. While co-occurrence and metabolic models describe potential ecological and metabolic relationships, the signaling-based layer of communication remains underexplored. In this project, we systematically identified and classified QS pathways from metagenome-assembled genomes (MAGs) using a KEGG-based reference framework comprising 41 pathways and 18 signal molecule types. We developed a sender-receiver inference model to quantify communication potentials and reconstructed directional QS networks that reveal how microbial taxa interact through signal synthesis and perception. This work establishes a methodological framework for integrating chemical signaling into microbial interaction modeling and provides novel ecological insights into microbial communication mechanisms in extreme environments.<br>The QS function annotation is being integrated into a published tool, <a target="_blank" rel="noopener" href="https://microbetag.readthedocs.io/en/latest/welcome.html">Microbetag</a>, designed for microbial network annotation (<a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-025-03769-2"><strong>ref</strong></a>). The work is in collaboration with Dr. Haris Zafeiropoulos under the supervision of Dr. Karoline Faust (KU Leuven, Belgium). </p>
<h2 id="Skills-and-Expertise"><a href="#Skills-and-Expertise" class="headerlink" title="Skills and Expertise"></a>Skills and Expertise</h2><p><em>Molecular biology skills</em></p>
<p>Peng is proficient in standard laboratory techniques (including DNA extraction, PCR, gel electrophoresis, aseptic technique, and microbial culturing).</p>
<p><em>Bioinformatic skills</em></p>
<p>Peng excels in comprehensive metagenomic analysis, with expertise spanning amplicon and shotgun sequencing. His contributions include developing and maintaining the Galaxy Platform for amplicon&#x2F;functional gene analysis (DMAP) and the Integrated Network Analysis Pipeline (iNAP), showcasing his skills in coding and bioinformatics infrastructure development. Proficient in the standard metagenomic analysis workflow, Peng adeptly utilizes quality control (Trimmomatic, NextPolish), read assembly (MEGAHIT, metaSPAdes, IDBA-UD, OPERA-MS), and genome binning software (metaWRAP, metaBAT2, MAXBIN2, concoct, VAMB), among others.</p>
<p><em>Computer skills</em></p>
<p>A self-proclaimed computer enthusiast and emerging geek, Peng boasts advanced skills in Linux (Shell&#x2F;Bash), Python, and R programming. He is keenly interested in AI technologies and prompt engineering, with practical experience in ChatGPT and MidJourney. His expertise extends to statistical visualization, where he excels in R, Cytoscape, and Gephi, illustrating his versatile computational abilities.</p>
<p>Peng is the creator and maintainer of the homepage of Professor Ye Deng’s lab (<a target="_blank" rel="noopener" href="https://denglab.org.cn/">MEM</a>).</p>


  

</article>



              </div>
            </div>
          </div>
        </div>
      </div>
    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">Search</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">Keyword</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <p>&copy; Copyright 2025 Xi Emmett Peng, All rights reserved.</p> <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">Blog works best with JavaScript enabled</div>
  </noscript>
</body>
</html>
