

<!DOCTYPE html>
<html lang="en" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">
  <link rel="apple-touch-icon" sizes="76x76" href="/img/page_avatar.png">
  <link rel="icon" href="/img/page_avatar.png">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="Emmett Peng">
  <meta name="keywords" content="">
  
    <meta name="description" content="最近在做一批宏基因组数据的拼接工作，这里的拼接主要指从Read到Contig这个水平。然后因为是第一次做，有很多没懂的地方，在学习和实践的过程中，总结了一下。写成几篇博客记录一下，以后又不清楚了可以回来看一下。这篇博客主要记录一下宏基因组拼接的算法之一——基于德布莱英图（De Bruijn graph）的方法。 一、为什么使用德布莱英图？对于短序列read的拼接有两种基本算法，一种是基于Overl">
<meta property="og:type" content="article">
<meta property="og:title" content="基于德布莱英图（De Bruijn graph）的宏基因组de novo拼接">
<meta property="og:url" content="https://emmettpeng.github.io/2024/02/26/de-bruijn-graph/index.html">
<meta property="og:site_name" content="Emmett&#39;s Note">
<meta property="og:description" content="最近在做一批宏基因组数据的拼接工作，这里的拼接主要指从Read到Contig这个水平。然后因为是第一次做，有很多没懂的地方，在学习和实践的过程中，总结了一下。写成几篇博客记录一下，以后又不清楚了可以回来看一下。这篇博客主要记录一下宏基因组拼接的算法之一——基于德布莱英图（De Bruijn graph）的方法。 一、为什么使用德布莱英图？对于短序列read的拼接有两种基本算法，一种是基于Overl">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://img-blog.csdnimg.cn/20201025161548591.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center">
<meta property="og:image" content="https://img-blog.csdnimg.cn/20201025161558590.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center">
<meta property="og:image" content="https://img-blog.csdnimg.cn/20201025161609372.png#pic_center">
<meta property="og:image" content="https://img-blog.csdnimg.cn/20201025161626796.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center">
<meta property="og:image" content="https://img-blog.csdnimg.cn/20201025161638631.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center">
<meta property="article:published_time" content="2024-02-26T07:28:04.000Z">
<meta property="article:modified_time" content="2024-02-28T03:21:57.784Z">
<meta property="article:author" content="Emmett Peng">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="https://img-blog.csdnimg.cn/20201025161548591.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center">
  
  
  
  <title>基于德布莱英图（De Bruijn graph）的宏基因组de novo拼接 - Emmett&#39;s Note</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/font_1749284_hj8rtnfg7um.css">



<link rel="stylesheet" href="//at.alicdn.com/t/font_1736178_lbnruvf0jn.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"emmettpeng.github.io","root":"/","version":"1.9.5-a","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":false,"follow_dnt":true,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"woyaola":null,"cnzz":null,"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  


  
<meta name="generator" content="Hexo 6.3.0"></head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>emmett&#39;s notebook</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/">
                <i class="iconfont icon-home-fill"></i>
                <span>Home</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/">
                <i class="iconfont icon-archive-fill"></i>
                <span>Archives</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/">
                <i class="iconfont icon-category-fill"></i>
                <span>Categories</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/">
                <i class="iconfont icon-tags-fill"></i>
                <span>Tags</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/">
                <i class="iconfont icon-user-fill"></i>
                <span>About</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('/img/default.png') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="基于德布莱英图（De Bruijn graph）的宏基因组de novo拼接"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2024-02-26 15:28" pubdate>
          February 26, 2024 pm
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          3.2k words
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          27 mins
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">基于德布莱英图（De Bruijn graph）的宏基因组de novo拼接</h1>
            
            
              <div class="markdown-body">
                
                <p>最近在做一批宏基因组数据的拼接工作，这里的拼接主要指从Read到Contig这个水平。然后因为是第一次做，有很多没懂的地方，在学习和实践的过程中，总结了一下。写成几篇博客记录一下，以后又不清楚了可以回来看一下。这篇博客主要记录一下宏基因组拼接的算法之一——基于德布莱英图（<em>De Bruijn</em> graph）的方法。</p>
<h2 id="一、为什么使用德布莱英图？"><a href="#一、为什么使用德布莱英图？" class="headerlink" title="一、为什么使用德布莱英图？"></a>一、为什么使用德布莱英图？</h2><p>对于短序列read的拼接有两种基本算法，一种是基于Overlap Graph的算法，另一种是基于德布莱英图的算法。大多数为了桑格测序（Sanger sequencing）得到的测序数据拼接的算法都是基于Overlap Graph的，最典型的就是Overlap-layout-consensus方法，简称为OLC方法，这个方法通过两两之间有overlap区域的reads来构图，每个Read作为图中的点，而拥有足够的overlap则是建立图中的连接的基础。</p>
<p><img src="https://img-blog.csdnimg.cn/20201025161548591.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center" srcset="/img/loading.gif" lazyload alt="在这里插入图片描述"></p>
<p>这种方式在处理序列数很有限，而且序列间overlap区域很明显的时候能取得最好的效果。一些对于二代测序数据设计的assembler使用了OLC方法，但是这种传统的方法计算量实在过大，就算是非常简单的物种的从头组装也需要上百万甚至上千万条序列，所以仍然使用OLC方法显然不太实际。而且二代测序数据的读长片段，现在来说250bp和300bp的读长都是属于新兴事物，想要在茫茫多的短序列中找到可供使用的overlap区域也是OLC方法面临的瓶颈之一。</p>
<p>由于OLC方法跟不上更短的Read长度和更大的Read条数，更多的为二代测序数据拼接的assembler都选择了基于德布莱英图方法的，本文中将其简称为DBG，指<em>de Bruijn</em> graph。DBG方法通过将序列打断成更小的长度为<em>k</em>的小片段（称为<strong>k-mer</strong>）来降低计算的规模。对于一条长度为<em>N</em>的序列从头到尾（比如从5’-端到3‘-端）依次取k-mer一共可以取到*<em>N-k+1</em>个k-mer，每两个相邻的k-mer之间都有长度为k-1的overlap区域。</p>
<p><img src="https://img-blog.csdnimg.cn/20201025161558590.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center" srcset="/img/loading.gif" lazyload alt="在这里插入图片描述"></p>
<p>通过将序列打断为k-mers，德布莱英图方法降低了大量短序列数据集中的数据冗余性，就好比上图中的这条序列在被切为3-mers后，GAT这个k-mer是重复出现的，并在德布莱英图中被反复利用，也就达到了降低了整体数据量和计算量的目的。对于某一特定的组装来说，<strong>最大的有效k-mer size取决于读长和error rate，同时如何选择k的取值也是决定了组装质量的重要条件（我现在也正在调查研究这一点）。</strong>实际上，在拼接前是有方法可以先看看合适的k值的，比如<strong>kmergenie</strong>这款软件就可以给出一个预测的k-mer值。</p>
<p>德布莱英图的另一个很优秀的特性就是在基因组中的一些重复片段会在德布莱英图中被打断，避免了一些错误的组装结果（当然不能很好地解决）。</p>
<p><em>第一节的内容和图片整理来自：<a target="_blank" rel="noopener" href="https://banana-slug.soe.ucsc.edu/_media/bioinformatic_tools:abyss_technote_illumina.pdf">https://banana-slug.soe.ucsc.edu/_media/bioinformatic_tools:abyss_technote_illumina.pdf</a></em></p>
<h2 id="二、从德布莱英图到拼接结果"><a href="#二、从德布莱英图到拼接结果" class="headerlink" title="二、从德布莱英图到拼接结果"></a>二、从德布莱英图到拼接结果</h2><p>对于某一组测序数据来说，通过相应的算法获得了德布莱英图，怎么将图转化为序列呢？首先我们必须知道，德布莱英图是具有<strong>欧拉图（Eulerian Graph）</strong>的性质的，也就是说我们需要寻找的通路（path）是一条遍历每个边至少一次的。当然我们也希望它是遍历每个点至少一次，但是<strong>哈密顿图（Hamiltonian Graph）</strong>作为世界难解的NP-Complete问题，还很难被归化成一个具有概括性的模型。</p>
<p>根据上面的欧拉通路性质，对于下面的这组德布莱英图，我们认为它的拼接结果应该是GATTACATTACAA，或者说图中的ACA-CAT-ATT这个分支环路至少要被遍历一次。</p>
<p><img src="https://img-blog.csdnimg.cn/20201025161609372.png#pic_center" srcset="/img/loading.gif" lazyload alt="在这里插入图片描述"></p>
<p>事实上，当序列被打断为短k-mer后，序列的模糊性就变高了，就比如上图中的分支环路到底遍历几次，或者像下图这样的：</p>
<p><img src="https://img-blog.csdnimg.cn/20201025161626796.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center" srcset="/img/loading.gif" lazyload alt="在这里插入图片描述"></p>
<p>好像有很多条通路都可以满足条件。所以在实际情况中，assembler返回的是一组子通路（substring）的集合，通过这些子通路就可以组装出多条通路。比如针对上面这张图，返回的子通路集合是这样的{ACTGA, GACC, GAGTG, GAATG}，其中第一个元素代表这个图的“入口”，第二个元素好代表这个图的“出口”，第三个元素代表的是上面蓝色的分支环路，第四个元素代表的是下面绿色的分支环路，通过按顺序选择分支就可以得到一条完整的路径。</p>
<p>通过拼接得到的contig是不能有分支节点的，也就是说我们想要得到的就是简单的单一路径，或者说单一序列。所以如果我们最终得到的contig是下图这样的，我们就会把它们按分支节点拆开，作为最终的contig结果。</p>
<p><img src="https://img-blog.csdnimg.cn/20201025161638631.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L0VtbWV0dF9CaW9pbmZv,size_16,color_FFFFFF,t_70#pic_center" srcset="/img/loading.gif" lazyload alt="在这里插入图片描述"></p>
<p>这一部分其实还有大佬讲的更详细一点，参考：<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/54466660%E3%80%82">https://zhuanlan.zhihu.com/p/54466660。</a></p>
<h2 id="三、几个值得关注的问题（之后边学习边写）"><a href="#三、几个值得关注的问题（之后边学习边写）" class="headerlink" title="三、几个值得关注的问题（之后边学习边写）"></a>三、几个值得关注的问题（之后边学习边写）</h2><p>以上就是我现在比较关心的从read到contig的这样一个拼接过程的简单流程，但是在<strong>实用阶段</strong>中还有如下几个问题：</p>
<ol>
<li>以上介绍的都是单k-mer实现序列拼接，但是现在的Assemblers普遍输入k-mer list，也就是说它会有多个k值进行迭代式的拼接。为什么要用multi-k-mer的方法？k值以及k-max、k-min该如何选择？</li>
<li>Assembler的参数该如何选择？或者说该如何针对不同的数据情况选择？</li>
<li>德布莱英图只是一个基础，不同的Assembler，比如Megahit、MetaSPAdes、IDBA-UD等都在DBG的方法上进行了自己的改良优化，它们的详细原理是什么（关系到它们的适用场景和参数选择）？</li>
<li>还有很多使用中遇到的问题……</li>
</ol>
<p>在学习的过程中写了这些笔记，整理出来发成一个文档集。之后应该还会写以上几个问题的学习笔记和几款Assembler的使用说明（弄清楚了咋整的之后……），stay tuned :)</p>
<h3 id="参考文献和学习资料："><a href="#参考文献和学习资料：" class="headerlink" title="参考文献和学习资料："></a>参考文献和学习资料：</h3><ol>
<li><p><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/54466660">https://zhuanlan.zhihu.com/p/54466660</a></p>
</li>
<li><p><a target="_blank" rel="noopener" href="http://www.cs.bilkent.edu.tr/~calkan/teaching/cs481/pdfslides/20-cs481-intro_to_assembly.pdf">http://www.cs.bilkent.edu.tr/~calkan/teaching/cs481/pdfslides/20-cs481-intro_to_assembly.pdf</a></p>
</li>
<li><p><a target="_blank" rel="noopener" href="http://compbio.charite.de/tl_files/groupmembers/robinson/genome-assembly-1.pdf">http://compbio.charite.de/tl_files/groupmembers/robinson/genome-assembly-1.pdf</a></p>
</li>
<li><p><a target="_blank" rel="noopener" href="http://compbio.charite.de/tl_files/groupmembers/robinson/genome-assembly-2.pdf">http://compbio.charite.de/tl_files/groupmembers/robinson/genome-assembly-2.pdf</a></p>
</li>
<li><p>Pasolli E, Asnicar F, Manara S, et al. Extensive Unexplored Human Microbiome Diversity Revealed by Over 150,000 Genomes from Metagenomes Spanning Age, Geography, and Lifestyle. Cell. 2019;176(3):649-662.e20. doi:10.1016&#x2F;j.cell.2019.01.001</p>
</li>
<li><p>Forouzan E, Shariati P, Mousavi Maleki MS, Karkhane AA, Yakhchali B. Practical evaluation of 11 de novo assemblers in metagenome assembly. J Microbiol Methods. 2018 Aug;151:99-105. doi: 10.1016&#x2F;j.mimet.2018.06.007. Epub 2018 Jun 25. PMID: 29953874.</p>
</li>
<li><p><a target="_blank" rel="noopener" href="https://banana-slug.soe.ucsc.edu/_media/bioinformatic_tools:abyss_technote_illumina.pdf">https://banana-slug.soe.ucsc.edu/_media/bioinformatic_tools:abyss_technote_illumina.pdf</a></p>
</li>
</ol>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/Bioinformatics-tool/" class="category-chain-item">Bioinformatics tool</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>基于德布莱英图（De Bruijn graph）的宏基因组de novo拼接</div>
      <div>https://emmettpeng.github.io/2024/02/26/de-bruijn-graph/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>Author</div>
          <div>Emmett Peng</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>Posted on</div>
          <div>February 26, 2024</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>Licensed under</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - Attribution">
                    <i class="iconfont icon-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/2024/02/28/megahit/" title="Megahit: 高效迅速的metagenome assembler">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">Megahit: 高效迅速的metagenome assembler</span>
                        <span class="visible-mobile">Previous</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2024/01/30/clumpify/" title="Clumpify.sh (from BBMap): Remove PCR Duplicates">
                        <span class="hidden-mobile">Clumpify.sh (from BBMap): Remove PCR Duplicates</span>
                        <span class="visible-mobile">Next</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
  
  
    <article id="comments" lazyload>
      
  <div id="valine"></div>
  <script type="text/javascript">
    Fluid.utils.loadComments('#valine', function() {
      Fluid.utils.createScript('https://lib.baomitu.com/valine/1.5.1/Valine.min.js', function() {
        var options = Object.assign(
          {"appId":"YeHkpbZBcgEvVDUhALlkAvpl-MdYXbMMI","appKey":"jtxg2UEr7vvpFAljpNrl6Y1V","path":"window.location.pathname","placeholder":null,"avatar":"retro","meta":["nick","mail","link"],"requiredFields":[],"pageSize":10,"lang":"zh-CN","highlight":false,"recordIP":false,"serverURLs":"https://yehkpbzb.api.lncldglobal.com","emojiCDN":null,"emojiMaps":null,"enableQQ":false},
          {
            el: "#valine",
            path: window.location.pathname
          }
        )
        new Valine(options);
        Fluid.utils.waitElementVisible('#valine .vcontent', () => {
          var imgSelector = '#valine .vcontent img:not(.vemoji)';
          Fluid.plugins.imageCaption(imgSelector);
          Fluid.plugins.fancyBox(imgSelector);
        })
      });
    });
  </script>
  <noscript>Please enable JavaScript to view the comments</noscript>


    </article>
  


          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>Table of Contents</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">Search</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">Keyword</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
  
  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/4.3.1/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">Blog works best with JavaScript enabled</div>
  </noscript>
</body>
</html>
